### PR TITLE
Code samples

### PR DESCRIPTION
**Pull request by thallgren**
This PR contains a set of commits, each focused on some examples that cover some specific area. The details are in the commit comments.

### CODE REVIEW

File: examples_test/config_test.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+			t.Fatalf("unexpected result %v", result)
+		}
+	})
+}
+
+// TestHelloWorld_semiHardWired uses the special "lookup_key" function MuxLookupKey. This function use the configuration
+// option LookupProvidersKey where it expects to find a slice of "lookup_key" functions to use.
+func TestConfig_semiHardWired(t *testing.T) {
+	// Create "global" options for this Hiera session (global as in "kept in the top-level context". An executable may
+	// have an unlimited set of such contexts and there are no global variables involved.
```
Comment (hlindberg) [inline examples_test/config_test.go]: This is hard to grasp - it mentions "global", "session" "context" and "executable". How are "global" options for a session different from 'non "global"' ?
Is this just saying that you can have as many independent sessions as you like?

File: examples_test/config_test.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+	hiera.DoWithParent(context.Background(), provider.MuxLookupKey, configOptions, func(c px.Context) {
+		result := hiera.Lookup(hiera.NewInvocation(c, nil, nil), `hello`, nil, nil)
+		if result == nil || `hello world` != result.String() {
+			t.Fatalf("unexpected result %v", result)
+		}
+	})
+}
+
+/*
+ The remaining in tests in this file use the ConfigLookupKey provider. This provider will consult the configuration
```
Comment (hlindberg) [inline examples_test/config_test.go]:
```suggestion
 The remaining tests in this file use the ConfigLookupKey provider. This provider will consult the configuration
```

File: examples_test/config_test.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+ options HieraRoot, HieraConfigFileName, and HieraConfig to determine the path of the configuration file. Use of
+ HieraConfig is mutually exclusive to use of HieraRoot and HieraConfigFileName.
```
Comment (hlindberg) [inline examples_test/config_test.go]:
```suggestion
 HieraConfig is mutually exclusive with HieraRoot and HieraConfigFileName.
```

File: examples_test/config_test.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+ HieraRoot will default to the current working directory
```
Comment (hlindberg) [inline examples_test/config_test.go]:
```suggestion
 HieraRoot will default to the current working directory.
```

File: examples_test/config_test.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+ HieraConfigFileName will default to "hiera.yaml"
```
Comment (hlindberg) [inline examples_test/config_test.go]:
```suggestion
 HieraConfigFileName will default to "hiera.yaml".
```

File: examples_test/config_test.go [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+ If the HieraRoot or HieraConfig are relative paths, they will be considered relative to the current directory.
+
+ The HieraConfigFileName must be relative to the HieraRoot (or the current directory).
```
Comment (hlindberg) [inline examples_test/config_test.go]: Ambiguous - must the file be a child of the HieraRoot (i.e. if it is a full path), or is this telling me that a relative path is relative to the HieraRoot ? or what?

File: examples_test/config_test.go [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+*/
+
+// TestHelloWorld_yamlConfig uses the special "lookup_key" function ConfigLookupKey and HieraRoot.
```
Comment (hlindberg) [inline examples_test/config_test.go]: Again "special" without explanation. Is it a "composite" or "aggregate"/"aggregating" function perhaps?

File: examples_test/config_test.go [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+func TestHelloWorld_yamlConfig(t *testing.T) {
+	configOptions := make(map[string]px.Value)
+	configOptions[hieraapi.HieraRoot] = types.WrapString(`testdata`)
+
+	// Initialize a Hiera session with the ConfigLookupKey as the top-level function and perform a lookup and
+	// the created configOptions.
```
Comment (hlindberg) [inline examples_test/config_test.go]: Why would I do this - I am not sure I at this point really understand what `ConfigLookupKey` does for me...

Comment (hlindberg) [inline examples_test/config_test.go]: Sentence is strange it says "and perform a lookup and the created configOptions" did it mean to say "using the created configOptions" - if so, that is pretty clear from the code...

File: examples_test/config_test.go [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+	hiera.DoWithParent(context.Background(), provider.ConfigLookupKey, configOptions, func(c px.Context) {
+		result := hiera.Lookup(hiera.NewInvocation(c, nil, nil), `hello`, nil, nil)
+		if result == nil || `yaml data says hello` != result.String() {
+			t.Fatalf("unexpected result %v", result)
+		}
+	})
+}
+
+// TestHelloWorld_explicitYamlConfig is similar to TestHelloWorld_yamlConfig but uses HieraConfig
+// option to explicitly define what file to use.
```
Comment (hlindberg) [inline examples_test/config_test.go]:
```suggestion
// option to explicitly define the file to use.
```

File: examples_test/config_test.go [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+func TestHelloWorld_explicitYamlConfig(t *testing.T) {
+	configOptions := make(map[string]px.Value)
+	configOptions[hieraapi.HieraConfig] = types.WrapString(`testdata/hiera.yaml`)
+
+	// Initialize a Hiera session with the ConfigLookupKey as the top-level function and perform a lookup and
+	// the created configOptions.
+	hiera.DoWithParent(context.Background(), provider.ConfigLookupKey, configOptions, func(c px.Context) {
```
Comment (hlindberg) [inline examples_test/config_test.go]: same comment on similar above

File: examples_test/config_test.go [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+		result := hiera.Lookup(hiera.NewInvocation(c, nil, nil), `hello`, nil, nil)
+		if result == nil || `yaml data says hello` != result.String() {
+			t.Fatalf("unexpected result %v", result)
+		}
+	})
+}
+
+// TestHelloWorld_explicitYamlConfigFile is similar to TestHelloWorld_yamlConfig but uses a combination of
+// HieraRoot and HieraConfigFileName option.
```
Comment (hlindberg) [inline examples_test/config_test.go]: to achieve what?

File: examples_test/config_test.go [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+// option LookupProvidersKey where it expects to find a slice of "lookup_key" functions to use.
+func TestConfig_semiHardWired(t *testing.T) {
+	// Create "global" options for this Hiera session (global as in "kept in the top-level context". An executable may
+	// have an unlimited set of such contexts and there are no global variables involved.
+	configOptions := make(map[string]px.Value)
+
+	// The LookupProvidersKey stores a go slice of Hiera "lookup_key" functions that serve as the top level functions.
+	configOptions[provider.LookupProvidersKey] = types.WrapRuntime([]hieraapi.LookupKey{sayHello})
+
+	// Initialize a Hiera session with the MuxLookupKey as the top-level function and perform a lookup and
```
Comment (hlindberg) [inline examples_test/config_test.go]: MuxLookupKey introduced - what is that, why would I be using that?

File: examples_test/config_test.go [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+	hiera.DoWithParent(context.Background(), provider.ConfigLookupKey, configOptions, func(c px.Context) {
+		result := hiera.Lookup(hiera.NewInvocation(c, nil, nil), `hello`, nil, nil)
+		if result == nil || `yaml special data says hello` != result.String() {
+			t.Fatalf("unexpected result %v", result)
+		}
+	})
+}
+
+// TestHelloWorld_yamlAndSemiHardWired uses the MuxLookupKey to inject two lookup_key functions. The ConfigLookupKey
+// that consults the yaml config and the sayHello.
```
Comment (hlindberg) [inline examples_test/config_test.go]: This provides a clue to MuxLookupKey - it seems to build an aggregate function having a hierarchy of looking in one than the other - assuming the example shows configured to have higher precedence maybe spell that out.

File: examples_test/config_test.go [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+func TestHelloWorld_yamlAndSemiHardWired(t *testing.T) {
+	configOptions := make(map[string]px.Value)
+
+	// The LookupProvidersKey stores a go slice of Hiera "lookup_key" functions that serve as the top level functions.
```
Comment (hlindberg) [inline examples_test/config_test.go]: What does "serving as top level functions" mean?

File: examples_test/config_test.go [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+	// Initialize a Hiera session with the MuxLookupKey as the top-level function and perform a lookup and
+	// the created configOptions.
+	hiera.DoWithParent(context.Background(), provider.MuxLookupKey, configOptions, func(c px.Context) {
+		// A lookup of just "hello" should hit the first provider, the ConfigLookupKey.
+		result := hiera.Lookup(hiera.NewInvocation(c, nil, nil), `hello`, nil, nil)
+		if result == nil || `yaml data says hello` != result.String() {
+			t.Fatalf("unexpected result %v", result)
+		}
+
+		// A lookup of "howdy" is not found using the yaml configuration, so it hits the second provider, the sayHello.
```
Comment (hlindberg) [inline examples_test/config_test.go]: This is hard to understand just reading the code seems I need to go on a hunt for files...

File: examples_test/merge_test.go [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+			t.Fatalf("unexpected result %v", result)
+		}
+
+		// obtain the full map and compare
+		result = hiera.Lookup(internal.NewInvocation(c, nil, nil), `m`, nil, opts)
+		if !px.Wrap(c, map[string]string{`a`: `first value of a`, `b`: `second value of b`, `c`: `first value of c`}).Equals(result, nil) {
+			t.Fatalf("unexpected result %v", result)
+		}
+	})
+}
```
Comment (hlindberg) [inline examples_test/merge_test.go]: Would like an example how I can separately lookup two keys `key_a` and `key_b` (both being hashes) and do a hiera equivalent deep merge of one into the other.

File: examples_test/scope_test.go [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+			t.Fatalf("unexpected result %v", result)
+		}
+
+		// The value of key "ipl_c" is "x = %{c.x}, y = %{c.y}"
+		result = hiera.Lookup(internal.NewInvocation(c, s, nil), `ipl_c`, nil, nil)
+		if result == nil || `x = 10, y = 20` != result.String() {
+			t.Fatalf("unexpected result %v", result)
+		}
+	})
+}
```
Comment (hlindberg) [inline examples_test/scope_test.go]: Would like example how to create a scope from a `map[string]interface{}`

File: examples_test/testdata/data/common.yaml [hunk 18]
```diff
@@ -0,0 +1,3 @@
+hello: yaml data says hello
+ipl_a: interpolate <%{a}>
+ipl_c: x = %{c.x}, y = %{c.y}
```
Comment (hlindberg) [inline examples_test/testdata/data/common.yaml]: missing end of line?

File: examples_test/scope_test.go (hunk 17 above)
Comment (thallgren) [reply examples_test/scope_test.go]: Well, this *is* an example of that given that the type is just a cast with methods added to it, i.e. you can spell things out and do:
```
s := scope(map[string]interface{}{`foo`: `bar`})
```
but that's go syntax, not hiera.

File: examples_test/config_test.go [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+// TestConfig_hardwired utilizes Hiera in the simplest way possible. No configuration file and no options. Just
+// a function performs a lookup of a key.
+func TestConfig_hardwired(t *testing.T) {
+	// Use the hiera.DoWithParent to initialize a Hiera context (session) with the sayHello as the top-level function and
+	// perform a lookup.
+	//
+	// The DoWithParent is meant to be called once and the created context can then be used for any number of lookups that
+	// uses the same configuration. The sessions life-cycle can be compared to the compilers life-cycle in puppet.
```
[the remark this answers is not in the document]
Comment (hlindberg) [reply examples_test/config_test.go]: doh, naturally (a package) - so the context/session/scope is really just inside of the function that is given in that call - i.e. there is no state inside of "hiera" that is preserved after that call...

File: examples_test/scope_test.go (hunk 17 above)
Comment (hlindberg) [reply examples_test/scope_test.go]: ok - this is me not knowing if the two were equivalent

File: examples_test/config_test.go [hunk 20]
```diff
@@ -33,15 +40,17 @@ func TestConfig_hardwired(t *testing.T) {
 	})
 }
 
-// TestHelloWorld_semiHardWired uses the special "lookup_key" function MuxLookupKey. This function use the configuration
-// option LookupProvidersKey where it expects to find a slice of "lookup_key" functions to use.
+// TestHelloWorld_semiHardWired uses the "lookup_key" function MuxLookupKey. This function use the configuration
+// option LookupKeyFunctions where it expects to find a slice of "lookup_key" functions to use. Those functions form
+// top level hierarchy that is configurable from code. Very useful if you for instance want to introduce different
+// lookup layers such as "global", "environment", and "module" or in other ways build a complex lookup hierarchy that
+// service that goes beyond what can be defined in the yaml configuration.
```
Comment (hlindberg) [inline examples_test/config_test.go]: aha, so could I use the Mux to create a slice with ConfigLookupKey based resulting functions - one per module and thereby get "data in modules" or is more needed? (I could loop naturally)

Comment (thallgren) [reply examples_test/config_test.go]: Almost. You would need a specialized `ConfigLookupKey` function to cater for path resolution of environment and module roots. Hiera doesn't know about those concepts at all right now. We need to invent new options to carry them.

**Pull request merged**